### ISSUE TITLE
MAINT: arffread.py -- redundant if-condition

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#11864]
Something is strange about the `elif "yy"` in the code here: https://github.com/scipy/scipy/blob/master/scipy/io/arff/arffread.py#L273-L278

Should it be replaced with an `else` or does an `in pattern` need to be added? As it reads right now, it is basically a weird way to do `elif True:`, which itself is a weird way to do `else`.

```python
    @staticmethod
    def _get_date_format(atrv):
        m = r_date.match(atrv)
        if m:
            pattern = m.group(1).strip()
            # convert time pattern from Java's SimpleDateFormat to C's format
            datetime_unit = None
            if "yyyy" in pattern:
                pattern = pattern.replace("yyyy", "%Y")
                datetime_unit = "Y"
            elif "yy":
                pattern = pattern.replace("yy", "%y")
                datetime_unit = "Y"
            if "MM" in pattern:
                pattern = pattern.replace("MM", "%m")
                datetime_unit = "M"
```

I found this by running `vulture . --min-confidence 100 --exclude benchmarks/`

The rest of the output is below--I added something similar to the NumPy CI months ago, so was just curious what would show up for us.

<details>

```
scipy/_lib/_util.py:387: unused variable 'exc_type' (100% confidence)
scipy/_lib/_util.py:387: unused variable 'exc_value' (100% confidence)
scipy/cluster/hierarchy.py:3420: unused variable 'mhr' (100% confidence)
scipy/integrate/_ode.py:1179: unused variable 'icomp' (100% confidence)
scipy/integrate/_ode.py:1179: unused variable 'nr' (100% confidence)
scipy/integrate/_ode.py:1179: unused variable 'xold' (100% confidence)
scipy/io/arff/arffread.py:276: redundant if-condition (100% confidence)
scipy/io/harwell_boeing/hb.py:213: unused variable 'right_hand_sides_nlines' (100% confidence)
scipy/linalg/tests/test_decomp.py:2131: unsatisfiable 'if' condition (100% confidence)
scipy/linalg/tests/test_interpolative.py:42: unsatisfiable 'if' condition (100% confidence)
scipy/optimize/optimize.py:1463: unused variable 'fp1' (100% confidence)
scipy/optimize/tests/test__basinhopping.py:99: unused variable 'accepted' (100% confidence)
scipy/optimize/tests/test__basinhopping.py:290: unused variable 'accepted' (100% confidence)
scipy/optimize/tests/test__basinhopping.py:298: unused variable 'accepted' (100% confidence)
scipy/sparse/bsr.py:623: unused variable 'in_shape' (100% confidence)
scipy/sparse/linalg/isolve/iterative.py:119: unused variable 'atol_default' (100% confidence)
scipy/stats/_continuous_distns.py:2709: unused variable 'g2gm12' (100% confidence)
setup.py:207: unused variable 'exception_type' (100% confidence)
setup.py:207: unused variable 'exception_value' (100% confidence)
tools/refguide_check.py:338: unused variable 'is_label' (100% confidence)
```
</details>

**Comment by distroinfinity** [comment]
Hello @tylerjereddy I would like to work on this issue and as I understand I just need to replace elif "yy" with something more suitable. I am new to open source and was hoping if you can guide me through this?

**Comment by tylerjereddy** [comment]
I went to the docs for java [`SimpleDateFormat`](https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html) and noticed that there are example datetime formats that do *not* include `yy` at all, so I suspect you'd want to change that to `elif "yy" in pattern:`.

We have a [contributor guide](https://docs.scipy.org/doc/scipy/reference/dev/contributor/contributor_toc.html) with some details on our workflow. The [GitHub PR docs](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) are pretty good too.

Normally for code fixes we'd also add a unit test that fails before the fix--in this case, that might be challenging though, since the behavior may not change much if at all. Definitely cool if you find such a case, but it may require looking at the code before/after the change to see what kind of data gets through.

**Comment by distroinfinity** [comment]
![Screenshot from 2020-04-17 21-37-37__01](https://user-images.githubusercontent.com/59890794/79591967-ea557280-80f6-11ea-9cac-9b5c9636a4b3.png)
Sir I'm getting a different version of scipy after setting up the development environment. Also it says the word 'dev' will be written in the version. Is it fine ? Have i installed scipy correctly ?

**Comment by rgommers** [comment]
Yes, you have done it correctly, you have installed the latest release.

**Comment by distroinfinity** [comment]
1.
I changed elif "yy" to elif "yy" in pattern,
and for testing I modified the "scipy/io/arff/tests/test_arffread.py" file by adding a new test case 
 def test_yeari_attribute(self):
        expected = np.array([
             '99-01-31',
            '04-12-01',
            '17-04-28',
            '00-09-10',
            '13-11-30',
            '31-10-15'
        ], dtype='datetime64[Y]')
        assert_array_equal(self.data["attr_yeari"], expected)

2. 

Also I tried adding a new "@ATTRIBUTE attr_yeari DATE yy-MM-dd" and the corresponfing values '99-01-31','04-12-01','17-04-28','00-09-10','13-11-30','31-10-15'
in scipy/io/arff/tests/data/test7.arff , to perform the test.

3.

When i performed the test by running "python runtests.py -v -t scipy/io/arff/tests/test_arffread.py" I am getting this error

TypeError: invalid type promotion

The above exception was the direct cause of the following exception:
scipy/io/arff/tests/test_arffread.py:212: in test_yeari_attribute
    assert_array_equal(self.data["attr_yeari"], expected)
E   DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
        expected   = array(['0099', '0004', '0017', '0000', '0013', '0031'],
      dtype='datetime64[Y]')
        self       = <scipy.io.arff.tests.test_arffread.TestDateAttribute object at 0x7fbd8d2a8eb0>

**Comment by tylerjereddy** [comment]
> I changed elif "yy" to elif "yy" in pattern,

Those look the same? You might want to open a WIP (work-in-progress) PR (pull request) so we can take a look at your code without trying to copy-paste from the description above.

**Comment by distroinfinity** [comment]
I submitted a Work In in Progress PR #11888. Please review it.

**Comment by onshek** [comment]
@tylerjereddy @distroinfinity Does anyone still work on this? I'm willing to have a try.

**Comment by tylerjereddy** [comment]
@onshek I think this proved a bit more complicated than initially projected, so that's part of the reason the cross-linked PR stalled.

**Comment by onshek** [comment]
> @onshek I think this proved a bit more complicated than initially projected, so that's part of the reason the cross-linked PR stalled.

Do you mean I should take care of the related tests?  I think I can handle that.
Any suggestions or thoughts will be appreciated.
